imap/imap_proxy.c

```diff
@@ -864,32 +865,210 @@ static char *editflags(char *flags)
+            if (myrights & ACL_READ) {
+                appenduid = strchr(s->last_result.s, '[');
+                /* skip over APPENDUID */
+                if (appenduid) {
+                    appenduid += strlen("[appenduid ");
+                    b = strchr(appenduid, ']');
+                    if (b) *b = '\0';
+                    prot_printf(imapd_out, "%s OK [COPYUID %s] %s\r\n", tag,
+                                appenduid, error_message(IMAP_OK_COMPLETED));
+                }
+                else
+                    prot_printf(imapd_out, "%s OK %s\r\n", tag, s->last_result.s);
+            }
+            else {
+                prot_printf(imapd_out, "%s OK %s\r\n", tag,
+                            error_message(IMAP_OK_COMPLETED));
+            }
+        } else {
+            prot_printf(imapd_out, "%s %s", tag, s->last_result.s);
+        }
+    } else {
+        /* abort the append */
+        prot_printf(s->out, " {0+}\r\n\r\n");
+        pipe_until_tag(s, tag, 0);
+
+        /* report failure */
+        prot_printf(imapd_out, "%s NO inter-server COPY failed\r\n", tag);
+    }
+
+    /* free dynamic memory */
+    free_hashu64_table(&crock.msg_byuid, NULL);
+    seqset_free(&crock.uidset);
+    free(freeme);
+}
+
+int proxy_fetch(char *sequence, int usinguid, unsigned items,
+                void (*item_cb)(uint32_t seqno, unsigned item,
+                                void *datap, void *rock),
+                void *rock)
+{
+    seqset_t *seqset = !usinguid ? seqset_parse(sequence, NULL, 0) : NULL;
+    struct buf data = BUF_INITIALIZER;
     char mytag[128];
-    struct d {
-        char *idate;
-        char *flags;
-        unsigned int seqno, uid;
-        struct d *next;
-    } *head, *p, *q;
+    char sep = '(';
     int c;
 
-    /* find out what the flags & internaldate for this message are */
     proxy_gentag(mytag, sizeof(mytag));
-    prot_printf(backend_current->out,
-                "%s %s %s (Flags Internaldate)\r\n",
-                tag, usinguid ? "Uid Fetch" : "Fetch", sequence);
-    head = (struct d *) xmalloc(sizeof(struct d));
-    head->flags = NULL; head->idate = NULL;
-    head->seqno = head->uid = 0;
-    head->next = NULL;
-    p = head;
-    /* read all the responses into the linked list */
+    prot_printf(backend_current->out, "%s %s %s ",
+                mytag, usinguid ? "Uid Fetch" : "Fetch", sequence);
+    if (items & FETCH_FLAGS) {
+        prot_printf(backend_current->out, "%c%s", sep, "Flags");
+        sep = ' ';
+    }
+    if (items & FETCH_INTERNALDATE) {
+        prot_printf(backend_current->out, "%c%s", sep, "Internaldate");
+        sep = ' ';
+    }
+    if (items & FETCH_RFC822) {
+        prot_printf(backend_current->out, "%c%s", sep, "Rfc822.peek");
+        sep = ' ';
+    }
+    if (items & FETCH_UID) {
+        prot_printf(backend_current->out, "%c%s", sep, "Uid");
+        sep = ' ';
+    }
+    prot_puts(backend_current->out, ")\r\n");
```

Comment:
I think if `items` was `0` here, then we're going to output a `)` without having output a `(` first.  Looks like the other places using this pattern do something like `if (sep == '(') prot_puts(out, "(");` just before outputting the `)` .

Reply:
I've added protection from this by bailing if `items == 0`